File: server/room-stays.js
// ===== Lịch sử ở phòng =====
// Trước đây chuyển phòng chỉ ĐÈ students.room_id, nên sau khi chuyển thì hệ thống không còn biết
// người đó từng ở phòng cũ. Hậu quả: tiền điện nửa tháng đầu của họ bị đổ hết sang đầu người ở lại.
// Mọi thay đổi phòng đều phải đi qua đây để giữ lại dấu vết.
//
// Quy ước ngày:
//   - Trả phòng ngày D  -> lượt ở kết thúc NGÀY D (vẫn tính trọn ngày D, khớp với check_out_date).
//   - Chuyển phòng ngày D -> lượt CŨ kết thúc ngày D-1, lượt MỚI bắt đầu ngày D
//     (ngày D là ngày đầu tiên ở phòng mới -> không bị tính điện ở cả 2 phòng cùng một ngày).

const { query } = require('./db');
const { addDays } = require('./billing');

const run = (db, sql, params) => (db && db.query ? db.query(sql, params) : query(sql, params));

// Lượt ở đang mở (chưa có ngày kết thúc) của một học viên
async function openStayOf(db, studentId) {
  const { rows } = await run(db, 'SELECT * FROM room_stays WHERE student_id=$1 AND to_date IS NULL ORDER BY from_date DESC LIMIT 1', [studentId]);
  return rows[0] || null;
}

// Đóng lượt đang mở tại ngày `toDate` (tính trọn ngày đó)
async function closeStay(db, studentId, toDate) {
  const cur = await openStayOf(db, studentId);
  if (!cur) return null;
  // Ngày kết thúc trước ngày bắt đầu = lượt ở này chưa từng xảy ra -> bỏ hẳn, đừng để lại dòng vô nghĩa
  if (toDate < String(cur.from_date).slice(0, 10)) {
    await run(db, 'DELETE FROM room_stays WHERE id=$1', [cur.id]);
    return null;
  }
  const { rows } = await run(db, 'UPDATE room_stays SET to_date=$1 WHERE id=$2 RETURNING *', [toDate, cur.id]);
  return rows[0];
}

// Mở lượt ở mới
async function openStay(db, studentId, roomId, fromDate) {
  if (!roomId || !fromDate) return null;
  const { rows } = await run(db, 'INSERT INTO room_stays (student_id, room_id, from_date, to_date) VALUES ($1,$2,$3,NULL) RETURNING *', [studentId, roomId, fromDate]);
  return rows[0];
}

// Vào ở / nhận phòng ngày `date`
async function checkIn(db, studentId, roomId, date) {
  await closeStay(db, studentId, addDays(date, -1)); // dọn lượt cũ còn treo (nếu có)
  return openStay(db, studentId, roomId, date);
}

// Chuyển sang phòng khác ngày `date`: lượt cũ hết ngày D-1, lượt mới từ ngày D
async function transfer(db, studentId, newRoomId, date) {
  await closeStay(db, studentId, addDays(date, -1));
  return openStay(db, studentId, newRoomId, date);
}

// Trả phòng hẳn ngày `date` (tính trọn ngày D)
async function checkOut(db, studentId, date) {
  return closeStay(db, studentId, date);
}

// Đồng bộ lượt ở đang mở với hồ sơ — dùng cho đường SỬA HỒ SƠ.
// Sửa ô "Phòng" trong form hồ sơ KHÔNG kèm ngày chuyển, nên không thể coi là chuyển phòng thật
// (chuyển thật phải đi qua chức năng "Chuyển phòng" để còn chốt chỉ số điện). Ở đây hiểu là SỬA NHẦM:
// đính chính lượt đang mở chứ không đẻ ra một lượt mới trong lịch sử.
async function reconcile(db, studentId, roomId, checkInDate, checkOutDate) {
  const cur = await openStayOf(db, studentId);
  const closed = checkOutDate || null;

  if (!roomId || !checkInDate) { // hồ sơ không còn phòng -> lượt đang mở là vô nghĩa
    if (cur) await run(db, 'DELETE FROM room_stays WHERE id=$1', [cur.id]);
    return null;
  }
  if (!cur) {
    const st = await openStay(db, studentId, roomId, checkInDate);
    if (closed) await closeStay(db, studentId, closed);
    return st;
  }

  // Có lịch sử chuyển phòng -> from_date của lượt đang mở là NGÀY CHUYỂN, không phải ngày nhận phòng.
  // Đè check_in_date lên đó sẽ phá lịch sử, nên chỉ đính chính phòng.
  const n = (await run(db, 'SELECT COUNT(*)::int c FROM room_stays WHERE student_id=$1', [studentId])).rows[0].c;
  if (n > 1) await run(db, 'UPDATE room_stays SET room_id=$1 WHERE id=$2', [roomId, cur.id]);
  else await run(db, 'UPDATE room_stays SET room_id=$1, from_date=$2 WHERE id=$3', [roomId, checkInDate, cur.id]);

  if (closed) await closeStay(db, studentId, closed);
  return openStayOf(db, studentId);
}

module.exports = { openStayOf, openStay, closeStay, checkIn, transfer, checkOut, reconcile };

File: internal/roomstays/roomstays.go
// Package roomstays — lịch sử ở phòng (nền tảng chia điện đúng). Port từ server/room-stays.js.
// Quy ước ngày: trả phòng ngày D -> lượt kết thúc ngày D; chuyển phòng ngày D -> lượt cũ hết D-1, mới từ D.
package roomstays

import (
	"context"
	"errors"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgtype"
	"ktx/internal/billing"
	"ktx/internal/db"
)

type Stay struct {
	ID       int
	RoomID   *int
	FromDate string
	ToDate   *string
}

// OpenStayOf: lượt ở đang mở (to_date NULL) của HV. server/room-stays.js:17-20
func OpenStayOf(ctx context.Context, q db.Querier, studentID int) (*Stay, error) {
	var s Stay
	var roomID *int
	var from, to pgtype.Date
	err := q.QueryRow(ctx, "SELECT id, room_id, from_date, to_date FROM room_stays WHERE student_id=$1 AND to_date IS NULL ORDER BY from_date DESC LIMIT 1", studentID).
		Scan(&s.ID, &roomID, &from, &to)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, nil
		}
		return nil, err
	}
	s.RoomID = roomID
	if from.Valid {
		s.FromDate = from.Time.Format("2006-01-02")
	}
	if to.Valid {
		t := to.Time.Format("2006-01-02")
		s.ToDate = &t
	}
	return &s, nil
}

// CloseStay: đóng lượt đang mở tại ngày toDate (tính trọn ngày đó). server/room-stays.js:23-33
func CloseStay(ctx context.Context, q db.Querier, studentID int, toDate string) error {
	cur, err := OpenStayOf(ctx, q, studentID)
	if err != nil || cur == nil {
		return err
	}
	// Kết thúc trước ngày bắt đầu = lượt chưa từng xảy ra -> xoá hẳn.
	if toDate < cur.FromDate {
		_, err = q.Exec(ctx, "DELETE FROM room_stays WHERE id=$1", cur.ID)
		return err
	}
	_, err = q.Exec(ctx, "UPDATE room_stays SET to_date=$1 WHERE id=$2", toDate, cur.ID)
	return err
}

// OpenStay: mở lượt ở mới. server/room-stays.js:36-40
func OpenStay(ctx context.Context, q db.Querier, studentID int, roomID *int, fromDate string) error {
	if roomID == nil || fromDate == "" {
		return nil
	}
	_, err := q.Exec(ctx, "INSERT INTO room_stays (student_id, room_id, from_date, to_date) VALUES ($1,$2,$3,NULL)", studentID, *roomID, fromDate)
	return err
}

// CheckIn: vào ở / nhận phòng ngày date. server/room-stays.js:43-46
func CheckIn(ctx context.Context, q db.Querier, studentID int, roomID *int, date string) error {
	if err := CloseStay(ctx, q, studentID, billing.AddDays(date, -1)); err != nil {
		return err
	}
	return OpenStay(ctx, q, studentID, roomID, date)
}

// Transfer: chuyển phòng ngày date (cũ hết D-1, mới từ D). server/room-stays.js:49-52
func Transfer(ctx context.Context, q db.Querier, studentID int, newRoomID *int, date string) error {
	if err := CloseStay(ctx, q, studentID, billing.AddDays(date, -1)); err != nil {
		return err
	}
	return OpenStay(ctx, q, studentID, newRoomID, date)
}

// CheckOut: trả phòng hẳn ngày date. server/room-stays.js:55-57
func CheckOut(ctx context.Context, q db.Querier, studentID int, date string) error {
	return CloseStay(ctx, q, studentID, date)
}

// Reconcile: đồng bộ lượt đang mở với hồ sơ (đường SỬA HỒ SƠ). server/room-stays.js:63-85
func Reconcile(ctx context.Context, q db.Querier, studentID int, roomID *int, checkInDate, checkOutDate string) error {
	cur, err := OpenStayOf(ctx, q, studentID)
	if err != nil {
		return err
	}
	if roomID == nil || checkInDate == "" {
		if cur != nil {
			_, err = q.Exec(ctx, "DELETE FROM room_stays WHERE id=$1", cur.ID)
		}
		return err
	}
	if cur == nil {
		if err := OpenStay(ctx, q, studentID, roomID, checkInDate); err != nil {
			return err
		}
		if checkOutDate != "" {
			return CloseStay(ctx, q, studentID, checkOutDate)
		}
		return nil
	}
	var n int
	if err := q.QueryRow(ctx, "SELECT COUNT(*)::int FROM room_stays WHERE student_id=$1", studentID).Scan(&n); err != nil {
		return err
	}
	if n > 1 {
		_, err = q.Exec(ctx, "UPDATE room_stays SET room_id=$1 WHERE id=$2", *roomID, cur.ID)
	} else {
		_, err = q.Exec(ctx, "UPDATE room_stays SET room_id=$1, from_date=$2 WHERE id=$3", *roomID, checkInDate, cur.ID)
	}
	if err != nil {
		return err
	}
	if checkOutDate != "" {
		return CloseStay(ctx, q, studentID, checkOutDate)
	}
	return nil
}
